### PR TITLE
Add tree dump utility to E2E test framework and fix Image border issue

### PR DESCRIPTION
**Pull request by ddalp**
- Created a TreeDump custom view manager and added it to E2E test project, so test can dump UIElement tree at runtime to verify the expected tree structure and layout properties. The TreeDump control can be configured to take two parameters: dumpID: specifies the dump file name to read from Assets for comprison or used to generate the .out file when match failed; uiaID: specify which node in the UIElement tree to start comparison, or the whole tree when leaving blank.
- Fix the border issue for Image #3456 , we need to put the Image canvas under a border element, so we can apply border properties on it. 
- Add E2E test for the image border properties, using the newly added TreeDump custom view.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/3754)

### CODE REVIEW

File: packages/E2ETest/windows/TreeDumpLibrary/VisualTreeDumper.cs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            {
+                var v = _translator.PropertyValueToString(propertyName, value);
+                if (_filter.ShouldVisitPropertyValue(v))
+                {
+                    _logger.LogProperty(_indent + 1, propertyName, v);
+                }
+            }
+        }
+
+        public static string DumpToXML(DependencyObject root, DependencyObject excludedNode)
```
Comment (licanhua) [inline packages/E2ETest/windows/TreeDumpLibrary/VisualTreeDumper.cs]: >DumpToXML [](start = 29, length = 9)

Please rename this. In the first design, the output is XML, but finally it's just a simple text with indent

File: packages/E2ETest/windows/TreeDumpLibrary/TreeDumpControlViewManager.cs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+            {
+                var bounds = ApplicationView.GetForCurrentView().VisibleBounds;
+                if (bounds.Width != 1024 || bounds.Height != 768)
+                {
+                    // Dump disabled when window size is not 1024x768!
+                    UpdateResult(false /*matchDump*/, "Window has been resized, dump comparison is only valid at default launch size: 1024x768!");
+                }
+                else
+                {
+                    await MatchTreeDumpFromLayoutUpdateAsync();
```
Comment (licanhua) [inline packages/E2ETest/windows/TreeDumpLibrary/TreeDumpControlViewManager.cs]: >MatchTreeDumpFromLayoutUpdateAsync [](start = 26, length = 34)

Will it introduce unnecessary layout cycle? For example. View is ready -> Layout updated -> not matched -> Textblocked updated -> Layout updated ->

File: packages/E2ETest/windows/ReactUWPTestApp/Assets/TreeDump/ImageWithBorder.treedump [hunk 3]
```diff
@@ -0,0 +1,16 @@
+[Windows.UI.Xaml.Controls.Border]
+    Padding=0,0,0,0
```
Comment (licanhua) [inline packages/E2ETest/windows/ReactUWPTestApp/Assets/TreeDump/ImageWithBorder.treedump]: Possible just make the extension name .treedump as .txt. customer doesn't need to setup a tool to open .treedump.

File: packages/E2ETest/windows/TreeDumpLibrary/VisualTreeDumper.cs [hunk 4]
```diff
@@ -0,0 +1,179 @@
+using System;
```
Comment (licanhua) [inline packages/E2ETest/windows/TreeDumpLibrary/VisualTreeDumper.cs]: copyright

File: packages/E2ETest/app/ImageTestPage.tsx [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    var previousImageBorderState = imageWithBorder;
+    setImageBorder(!previousImageBorderState);
+ }
+  return(
+  <View style={styles.container}>
+    <Image
+      testID={IMAGE_ELEMENT}
+      style={imageWithBorder?styles.imageWithBorder:styles.image}
+      resizeMode={'center'}
+      source={{uri: 'http://facebook.github.io/react-native/img/header_logo.png'}}
```
Comment (licanhua) [inline packages/E2ETest/app/ImageTestPage.tsx]: >http://facebook.github.io/react-native/img/header_logo.png [](start = 21, length = 58)

make it to Assert. it may introduce unnecessary delay and flakiness of testing

File: vnext/ReactUWP/Views/Image/ImageViewManager.cpp [hunk 6]
```diff
@@ -103,15 +107,19 @@ const char *ImageViewManager::GetName() const {
 }
 
 XamlView ImageViewManager::CreateViewCore(int64_t tag) {
-  return ReactImage::Create().as<winrt::Canvas>();
+  auto image = ReactImage::Create().as<winrt::Canvas>();
+  auto border = winrt::Border();
```
Comment (kmelmon) [inline vnext/ReactUWP/Views/Image/ImageViewManager.cpp]: This introduces an extra XAML element in the tree that may be unnecessary.  Grid and StackPanel both allow rounded corners.  Can you switch the Canvas to one of those instead?  This would save an extra element.

Also - we should consider pushing the rounded corners down to Panel in lifted XAML.  Feel free to discuss/open a feature request on WinUI.

File: vnext/ReactUWP/Utils/PropertyUtils.h [hunk 7]
```diff
@@ -174,26 +174,47 @@ bool TryUpdateBorderProperties(
       UpdateControlBorderResourceBrushes(element, nullptr);
     }
   } else if (propertyName == "borderLeftWidth") {
-    if (propertyValue.isNumber())
+    if (propertyValue.isNumber()) {
       SetBorderThickness(node, element, ShadowEdges::Left, propertyValue.asDouble());
+    } else if (propertyValue.isNull()) {
+      SetBorderThickness(node, element, ShadowEdges::Left, 0);
```
Comment (kmelmon) [inline vnext/ReactUWP/Utils/PropertyUtils.h]: Nice catch.  Is there a test for this scenario?

File: packages/E2ETest/wdio/test/Image.spec.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  HomePage.clickAndGotoImagePage();
+});
+
+describe('ImageWithoutBorderTest', () => {
+  it('ImageWithoutBorderTest Success', () => {
+    const result = ImageTestPage.getTreeDumpResult();
+    assert.ok(
+      result.includes('TreeDump:Passed'),
+      'Dump comparison passed for image without border!'
+    );
```
Comment (kmelmon) [inline packages/E2ETest/wdio/test/Image.spec.ts]: If the test fails, will anything be visible in the test output or as test artifacts other than this message?  This will be particularly important when the test is running in the CI loop.  Note this doesn't have to be done for this checkin - feel free to file an issue to improve this.

File: packages/E2ETest/app/ImageTestPage.tsx (hunk 5 above)
Comment (kmelmon) [inline packages/E2ETest/app/ImageTestPage.tsx]: >http://facebook.github.io/react-native/img/header_logo.png [](start = 21, length = 58)

I suggest packaging an image into the local bundle but that may require fixing #2452 first.  If so please leave a note in that bug to come back and fix this test.

File: packages/E2ETest/windows/TreeDumpLibrary/TreeDumpControlViewManager.cs (hunk 2 above)
Comment (ddalp) [reply packages/E2ETest/windows/TreeDumpLibrary/TreeDumpControlViewManager.cs]: It will cause layout once when text changes from Failed to Passed. If there are additional failure, I don't update TextBlock.Text. User has to click on it to view all saved error messages.

---
In reply to: [356297384](https://github.com/microsoft/react-native-windows/pull/3754#discussion_r356297384) [](ancestors = 356297384)

File: packages/E2ETest/windows/ReactUWPTestApp/Assets/TreeDump/ImageWithBorder.treedump (hunk 3 above)
Comment (ddalp) [reply packages/E2ETest/windows/ReactUWPTestApp/Assets/TreeDump/ImageWithBorder.treedump]: Will do.

---
In reply to: [356300132](https://github.com/microsoft/react-native-windows/pull/3754#discussion_r356300132) [](ancestors = 356300132)

File: vnext/ReactUWP/Utils/PropertyUtils.h (hunk 7 above)
Comment (ddalp) [reply vnext/ReactUWP/Utils/PropertyUtils.h]: I have this image test to test three scenarios: no border, switch to border and then switch back to no border. The third scenario was where I found this cleanup code missing.

---
In reply to: [356349717](https://github.com/microsoft/react-native-windows/pull/3754#discussion_r356349717) [](ancestors = 356349717)

File: packages/E2ETest/app/ImageTestPage.tsx (hunk 5 above)
Comment (ddalp) [reply packages/E2ETest/app/ImageTestPage.tsx]: Switching to using the embedded image data directly.

---
In reply to: [356352744](https://github.com/microsoft/react-native-windows/pull/3754#discussion_r356352744) [](ancestors = 356352744)

File: vnext/ReactUWP/Views/Image/ImageViewManager.cpp (hunk 6 above)
Comment (ddalp) [reply vnext/ReactUWP/Views/Image/ImageViewManager.cpp]: On top of the border issue, we also need to solve the problem for background on image, currently image brush is set to the image canvas's background so we need an additional container element to be able to draw background. Having extra border element solves both issues. Agree we should consider rounded corners for panel, we have similar issue for ViewViewManager too. 

---
In reply to: [356349285](https://github.com/microsoft/react-native-windows/pull/3754#discussion_r356349285) [](ancestors = 356349285)

File: vnext/ReactUWP/Views/Image/ImageViewManager.cpp [hunk 9]
```diff
@@ -103,15 +107,19 @@ const char *ImageViewManager::GetName() const {
...
+  border.Child(image);
```
Comment (kmelmon) [inline vnext/ReactUWP/Views/Image/ImageViewManager.cpp]: Yesterday in our discussion, I had a question about whether or not the image is supposed to draw underneath the border (as opposed to being inset).  Did you find out how this behaves on other platforms?

File: packages/E2ETest/app/ImageTestPage.tsx [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    height: 150,
+    width: 500,
+    margin: 10,
+  },
+});
+
+export function ImageTestPage() {
+  const [imageWithBorder, setImageBorder] = useState(false);
+  const [clickCount, setClickCount] = useState(0);
+  const onOressBorder = () => {
```
Comment (chrisglein) [inline packages/E2ETest/app/ImageTestPage.tsx]:
```suggestion
  const onPressBorder = () => {
```

File: packages/E2ETest/app/ImageTestPage.tsx [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+  <View>
+    <View testID={IMAGE_CONTAINER} style={imageWithBorder?styles.containerWithBorder:styles.container}>
+      <Image
+        style={imageWithBorder?styles.imageWithBorder:styles.image}
+        resizeMode={'center'}
+        source={{uri: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADMAAAAzCAYAAAA6oTAqAAAAEXRFWHRTb2Z0d2FyZQBwbmdjcnVzaEB1SfMAAABQSURBVGje7dSxCQBACARB+2/ab8BEeQNhFi6WSYzYLYudDQYGBgYGBgYGBgYGBgYGBgZmcvDqYGBgmhivGQYGBgYGBgYGBgYGBgYGBgbmQw+P/eMrC5UTVAAAAABJRU5ErkJggg=='}}        
+      />      
+    </View >
+    <Button title= {imageWithBorder?"Hide Border":"Show Border"} 
+          onPress={onOressBorder}
```
Comment (chrisglein) [inline packages/E2ETest/app/ImageTestPage.tsx]:
```suggestion
          onPress={onPressBorder} 
```

File: packages/E2ETest/wdio/pages/BasePage.ts [hunk 12]
```diff
@@ -23,6 +33,28 @@ export class BasePage {
...
+          '####Waiting for treedump comparison result ' +
+            waitCount +
+            '/' +
+            maxWait +
+            '...####'
+        );
+        wait(100);
+        waitCount += 1;
+      }
+    } while (waitCount <= maxWait && testResult == false);
```
Comment (chrisglein) [inline packages/E2ETest/wdio/pages/BasePage.ts]:
```suggestion
    } while (waitCount <= maxWait && !testResult);
```

File: packages/E2ETest/wdio/pages/BasePage.ts [hunk 13]
```diff
@@ -23,6 +33,28 @@ export class BasePage {
...
+      testResult = this.treeDumpResult.getText() == 'TreeDump:Passed';
+      if (!testResult) {
+        console.log(
+          '####Waiting for treedump comparison result ' +
+            waitCount +
+            '/' +
+            maxWait +
+            '...####'
+        );
+        wait(100);
```
Comment (chrisglein) [inline packages/E2ETest/wdio/pages/BasePage.ts]: Is there no event you can wait for?

File: packages/E2ETest/windows/ReactUWPTestApp/App.xaml.cs [hunk 14]
```diff
@@ -54,6 +58,7 @@ protected override void OnLaunched(LaunchActivatedEventArgs e)
         {
             base.OnLaunched(e);
             SystemNavigationManager.GetForCurrentView().AppViewBackButtonVisibility = AppViewBackButtonVisibility.Collapsed;
+            ApplicationView.GetForCurrentView().TryResizeView(new Size(1024, 768));
```
Comment (chrisglein) [inline packages/E2ETest/windows/ReactUWPTestApp/App.xaml.cs]: You had mentioned issues having consistent layout sizes. Did this work?

File: vnext/ReactUWP/Utils/PropertyUtils.h [hunk 15]
```diff
@@ -174,26 +174,47 @@ bool TryUpdateBorderProperties(
...
+    }
   } else if (propertyName == "borderTopWidth") {
-    if (propertyValue.isNumber())
+    if (propertyValue.isNumber()) {
       SetBorderThickness(node, element, ShadowEdges::Top, propertyValue.asDouble());
+    } else if (propertyValue.isNull()) {
+      SetBorderThickness(node, element, ShadowEdges::Top, 0);
+    }
```
Comment (chrisglein) [inline vnext/ReactUWP/Utils/PropertyUtils.h]: Could replace all of this with something like this:
```
std::optional<ShadowEdges> edge;
if (propertyName == "borderTopWidth") {
  edge = ShadowEdges::Top
} else if (propertyName == "borderRightWidth") {
```
and at the end
```
if (edge.has_value()) {
  if (propertyValue.isNumber()) {
    SetBorderThickness(node, element, edge.value(), propertyValue.asDouble());
  } else if (propertyValue.isNull()) {
    SetBorderThickness(node, element, edge.value(), 0);
  }
}
```
And/or put the mapping of string -> edge into a map.

**Comment by ddalp** [comment]
> CodeFlow for GitHub does not download executable files. If you trust this file and would like to run it, please download it yourself through GitHub.

Remove and git ignore this file

---
Refers to: packages/E2ETest/msbuild.binlog:1 in b6aedd8. [](commit_id = b6aedd8a9cd07d651582ed107fb29a3c7819d5cd, deletion_comment = False)

File: packages/E2ETest/wdio/test/Image.spec.ts (hunk 8 above)
Comment (ddalp) [reply packages/E2ETest/wdio/test/Image.spec.ts]: Added steps to generate artifacts for the generated .out file and .err file during the test.

---
In reply to: [356351609](https://github.com/microsoft/react-native-windows/pull/3754#discussion_r356351609) [](ancestors = 356351609)

File: vnext/ReactUWP/Views/Image/ImageViewManager.cpp (hunk 9 above)
Comment (ddalp) [reply vnext/ReactUWP/Views/Image/ImageViewManager.cpp]: Had a discussion with Keith offline, we decided to switch image from Canvas to Grid so we can set border properties on it and still just have 1 element for image in the tree. That means we can't support background property on Image for now, that requires at least three elements, and implementing TransferProperties, Reparent etc. Out of scope for this bug fix but I will open a new issue to track the background for image for fix at later time.

---
In reply to: [357410515](https://github.com/microsoft/react-native-windows/pull/3754#discussion_r357410515) [](ancestors = 357410515)

File: packages/E2ETest/wdio/pages/BasePage.ts (hunk 13 above)
Comment (ddalp) [reply packages/E2ETest/wdio/pages/BasePage.ts]: Was going to add event and wait for that, but we have issues with the optimized constructor with Context parameter on release build issue, so I decided to do polling for now. Will open a issue to revisit later after Jon fixes the Custom ViewManager bug.

---
In reply to: [357846482](https://github.com/microsoft/react-native-windows/pull/3754#discussion_r357846482) [](ancestors = 357846482)

File: packages/E2ETest/windows/ReactUWPTestApp/App.xaml.cs (hunk 14 above)
Comment (ddalp) [reply packages/E2ETest/windows/ReactUWPTestApp/App.xaml.cs]: Yes, by default we will launch the app at 800x600 and I also added the code to fail and don't do comparison if size is not 800x600 in tree dump too.

---
In reply to: [357847530](https://github.com/microsoft/react-native-windows/pull/3754#discussion_r357847530) [](ancestors = 357847530)

File: packages/E2ETest/windows/TreeDumpLibrary/TreeDumpControlViewManager.cs [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+
+                StorageFolder storageFolder = ApplicationData.Current.LocalFolder;
+                string fileNameError = "TreeDump\\" + m_dumpID + ".err";
+                try
+                {
+                    StorageFile errFile = await storageFolder.CreateFileAsync(fileNameError, CreationCollisionOption.ReplaceExisting);
+                    await Windows.Storage.FileIO.WriteTextAsync(errFile, m_errString);
+                }
+                catch (IOException)
+                {
```
[the remark this answers is not in the document]
Comment (ddalp) [reply packages/E2ETest/windows/TreeDumpLibrary/TreeDumpControlViewManager.cs]: But we are inside writing out the error text file, if that fails (i.e. file is opened in Notepad) I don't know where else to log...:)

---
In reply to: [357860121](https://github.com/microsoft/react-native-windows/pull/3754#discussion_r357860121) [](ancestors = 357860121)

File: vnext/ReactUWP/Utils/PropertyUtils.h (hunk 15 above)
Comment (ddalp) [reply vnext/ReactUWP/Utils/PropertyUtils.h]: Updated with map, thanks for the tips on std::optional..

---
In reply to: [357864798](https://github.com/microsoft/react-native-windows/pull/3754#discussion_r357864798) [](ancestors = 357864798)

File: vnext/ReactUWP/Utils/PropertyUtils.h [hunk 17]
```diff
@@ -32,6 +32,16 @@ static double DefaultOrOverride(double defaultValue, double x) {
   return x != c_UndefinedEdge ? x : defaultValue;
 };
 
+static const std::unordered_map<std::string, ShadowEdges> edgeTypeMap = {
```
Comment (licanhua) [inline vnext/ReactUWP/Utils/PropertyUtils.h]: >edgeTypeMap [](start = 58, length = 11)

If you are going to with map, I would suggest you put it in another PR. we have a lot of functions which can follow the same pattern (string -> enum) like padding

File: packages/E2ETest/app/ImageTestPage.tsx [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+      <Image
+        style={imageWithBorder?styles.imageWithBorder:styles.image}
+        resizeMode={'center'}
+        source={{uri: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADMAAAAzCAYAAAA6oTAqAAAAEXRFWHRTb2Z0d2FyZQBwbmdjcnVzaEB1SfMAAABQSURBVGje7dSxCQBACARB+2/ab8BEeQNhFi6WSYzYLYudDQYGBgYGBgYGBgYGBgYGBgZmcvDqYGBgmhivGQYGBgYGBgYGBgYGBgYGBgbmQw+P/eMrC5UTVAAAAABJRU5ErkJggg=='}}        
+      />      
+    </View >
+    <Button title= {imageWithBorder?"Hide Border":"Show Border"} 
+          onPress={onPressBorder} 
+          testID={SHOW_IMAGE_BORDER}/> 
+    <TreeDumpControl style={styles.treeDumpControl} dumpID={imageWithBorder?'ImageWithBorder':'ImageWithoutBorder'+clickCount} uiaID={IMAGE_CONTAINER} testID={TREE_DUMP_RESULT} />
```
Comment (licanhua) [inline packages/E2ETest/app/ImageTestPage.tsx]: >'ImageWithoutBorder'+clickCount [](start = 94, length = 31)

ImageWithoutBorder0.txt and ImageWithoutBorder2.txt is not readable, also you created a hard binding between ImageTestPage and Image.spec.ts.  If you refactor the test case like change the order or add a new click, it doesn't work anymore.

Comment (ddalp) [reply packages/E2ETest/app/ImageTestPage.tsx]: I am going to refactor the dumpID to "ImageWithoutBorder" and "ImageWithoutBorder-Subsequent", so it would work for click count > 2. The dump file is different if you initially don't have a border, or had a border then removed the border.

---
In reply to: [358381794](https://github.com/microsoft/react-native-windows/pull/3754#discussion_r358381794) [](ancestors = 358381794)

File: vnext/ReactUWP/Utils/PropertyUtils.h (hunk 17 above)
Comment (ddalp) [reply vnext/ReactUWP/Utils/PropertyUtils.h]: I'd like to keep this change as for now, we can open a new issue to track creating consolidated property mapping utility as we have many property mappings code in various shadow node/view managers.

---
In reply to: [358372703](https://github.com/microsoft/react-native-windows/pull/3754#discussion_r358372703) [](ancestors = 358372703)

File: packages/E2ETest/windows/TreeDumpLibrary/TreeDumpControlViewManager.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (ddalp) [reply packages/E2ETest/windows/TreeDumpLibrary/TreeDumpControlViewManager.cs]: Looks like I should use GenerateUniqueName collision option to avoid issues when file is currently opened already.

---
In reply to: [357875080](https://github.com/microsoft/react-native-windows/pull/3754#discussion_r357875080) [](ancestors = 357875080,357860121)

File: packages/E2ETest/wdio/pages/BasePage.ts (hunk 13 above)
Comment (kmelmon) [reply packages/E2ETest/wdio/pages/BasePage.ts]: Feel free to open an issue to add a WaitForIdle or similar.

---
In reply to: [357870760](https://github.com/microsoft/react-native-windows/pull/3754#discussion_r357870760) [](ancestors = 357870760,357846482)

Comment (ddalp) [reply packages/E2ETest/wdio/pages/BasePage.ts]: Create #3779  to track this.

---
In reply to: [358440736](https://github.com/microsoft/react-native-windows/pull/3754#discussion_r358440736) [](ancestors = 358440736,357870760,357846482)
